server/lib/auth.js

```diff
@@ -93,16 +95,53 @@ export function verifyTwoFactorAuthenticatorCode(encryptedTwoFactorAuthToken, tw
   return verified;
 }
 
-export function enforceTwoFactorAuthenticationOnPayouts(req, twoFactorAuthenticatorCode) {
+const getCacheKey = userId => {
+  return `${userId}_2fa_payment_limit`;
+};
+
+export async function rollingPayoutLimitTwoFactorAuthentication(
+  req,
+  twoFactorAuthenticatorCode,
+  hostRollingLimit,
+  expenseAmount,
+) {
   if (req.remoteUser.twoFactorAuthToken !== null) {
+    // 1. we check the 'cache' if the key exists: cacheKey=${user.id}_2fa_payment_limit
+    const userId = req.remoteUser.id;
+    const cacheKey = getCacheKey(userId);
+    const keyInCache = await cache.get(cacheKey);
+
     if (twoFactorAuthenticatorCode) {
       const verified = verifyTwoFactorAuthenticatorCode(req.remoteUser.twoFactorAuthToken, twoFactorAuthenticatorCode);
       if (!verified) {
         throw new Error('Two-factor authentication failed: invalid code. Please try again.');
       }
-      return;
+      // With 2FA code
+      // 2. if limit key exists, reset the limit; if limit key doesn't exist, initialise the limit
+      return {
+        cacheKey,
+        cacheValue: 0,
```

Comment:
Shouldn't that be `expenseAmount`? Otherwise the cache will always be at `0` when you pay the first expense

```suggestion
        cacheValue: expenseAmount,
```

Reply:
That's my understanding of what should happen from @znarf's details about how to write this: https://github.com/opencollective/opencollective/issues/3516

"allow payment and allow a follow up budget of $10,0000"

Reply:
I'm not following. 

If the user enter a correct 2FA code, then the limit should be initialized to `hostRollingLimit`. Do I get something wrong?

Reply:
Scratch my previous comment. I was confused by the meaning of `cacheValue` (this is why we should have a better name for it).

Because it's containing the sum of all expenses processed, setting it to `0` make sense to me.

Reply:
But the value returned by this helper is directly passed to `cache.set(cacheKey, cacheValue, CACHE_VALIDITY);`. So that means if I pay an expense for $500 while providing a valid 2FA code, the state of the cache after paying the expense will be $0.

Reply:
Bigger problem: if I try to pay a $500 expense while providing a valid 2FA code, then it fails, my cache state will be -$500

Reply:
I guess we need to be more precise in the specification and expectations. To make sure the code is doing what we actually want.

Reply:
> Bigger problem: if I try to pay a $500 expense while providing a valid 2FA code, then it fails, my cache state will be -$500

Yes, that's why I initially wrote the code so that the cache was not updated until *after* the payment was processed. Then we did all the cache locking. Then we discarded that.

So now based on the fact that we aren't going to implement cache locking now, we update the cache immediately and in the case the payment fails, we reset it to what it was before: https://github.com/opencollective/opencollective-api/blob/4b5ef9f6cbcb677e62c96ffbf7c4d6d72449e6e9/server/graphql/v1/mutations/expenses.js#L788

Reply:
> But the value returned by this helper is directly passed to cache.set(cacheKey, cacheValue, CACHE_VALIDITY);. So that means if I pay an expense for $500 while providing a valid 2FA code, the state of the cache after paying the expense will be $0.

> I guess we need to be more precise in the specification and expectations. To make sure the code is doing what we actually want.

Ok my understanding (again from @znarf's comment in the issue linked above) is that, given 4 payments to process that are $100, $200, $250, $300:

* I log in to make payments as a host admin. The host collective rolling limit is $500.
* I go to make my first payment, which is $100. I am asked for my 2FA code, and I enter it. The cache is created with a value of $0 out of $500 and the payment is processed. This first payment is **not** included in the limit.
* I make payment 2 of $200, the cache is now at $200 out of $500, the payment is processed.
* I make payment 3 of $250, the cache is now at $450 out of $500, the payment is processed.
* I make payment 4 of $300, the cache is now at $750 out of $500 - so we ask for 2FA again. I enter the code, the cache is reset to $0 out of $500, the payment is processed, etc

If this isn't everyone's understanding then can we use that example above and describe how you would expect it to work?

Reply:
Thank you for taking the time to write this example, the behavior it describes looks good to me.

Reply:
No worries, just wanted to make sure we were on the same page! I know that it could work a few different ways.